### DISCUSSION TITLE
Remove dependencies that rely on CombineSchedulers

### DISCUSSION THREAD

**Discussion by p4checo** [pointfreeco/swift-dependencies#16] [General]
Hi @stephencelis @mbrandonw 👋🏼 

I've been bringing the [ReactiveSwift fork of TCA](https://github.com/ReactiveCocoa/reactiveswift-composable-architecture) up to date with the latest upstream TCA changes, and while I think it's awesome that Dependencies are now on its own repo, the fact that `MainQueue` (and `MainRunLoop`**) has been included (and thus CombineSchedulers) is causing some issues when integrating with the ReactiveSwift fork. 

The problem arises because ReactiveSwift uses its own scheduler types and protocols which are not compatible with CombineScheduler's (and Combine's). While this is isn't a problem in Linux due to the `#if canInport(Combine)`, it does create a conflict in Apple platforms which isn't easy to circumvent.

My suggestion is then to move `MainQueue` (and `MainRunLoop`**) back to the TCA repo as "internal" dependencies, effectively removing the dependency on CombineSchedulers from Dependencies. I also assume Clocks are the way forward so I expect the reliance on schedulers to be temporary anyway.

What do you think? Thanks! 🙏🏼 

Notes:

** We aren't using `MainRunLoop` in the ReactiveSwift fork, but for the sake of consistency it would make sense to migrate it as well since it also depends on CombienSchedulers.

**Comment by mbrandonw** [comment]
Hi @p4checo, unfortunately we cannot do that because it would be a breaking change for anyone using the dependencies library. Also, beyond that, it's a goal of this library to provide a base set of dependencies that nearly every application will need to make use of, such as schedulers.

Can you describe more what is the exact problem you are having? Can you not disambiguate the `Scheduler` type by doing `Combine.Scheduler` vs `ReactiveSwift.Scheduler`?

And if this problem really is insurmountable without removing Combine schedulers, then you may have to fork this library to unblock you.

**Comment by p4checo** [comment]
> unfortunately we cannot do that because it would be a breaking change for anyone using the dependencies library. 
> Also, beyond that, it's a goal of this library to provide a base set of dependencies that nearly every application will need to make use of, such as schedulers.

I was afraid you would say that, but I understand 🥲 

> Can you describe more what is the exact problem you are having? Can you not disambiguate the Scheduler type by doing Combine.Scheduler vs ReactiveSwift.Scheduler?

Sure, apologies for not being clear from the start. Yes, we can easily disambiguate as you mention but the problem is that in the ReactiveSwift fork of TCA we've been defining the `\.mainQueue` dependency as a direct "translation" of the upstream TCA counterpart:

```swift
extension DependencyValues {
  public var mainQueue: DateScheduler {
    get { self[MainQueueKey.self] }
    set { self[MainQueueKey.self] = newValue }
  }

  private enum MainQueueKey: DependencyKey {
    static let liveValue = QueueScheduler.main as DateScheduler
    static let testValue = UnimplementedScheduler() as DateScheduler
  }
}
```

However with the new standalone dependencies by default it will now use the one defined as `AnySchedulerOf<DispatchQueue>` in [MainQueue.swift](https://github.com/pointfreeco/swift-dependencies/blob/main/Sources/Dependencies/DependencyValues/MainQueue.swift) causing builds to break, e.g.:

```swift
  final class ComposableArchitectureTests: XCTestCase {
    func testScheduling() async {
      struct Counter: ReducerProtocol {
        typealias State = Int
        enum Action: Equatable {
          case incrAndSquareLater
          case incrNow
          case squareNow
        }
        @Dependency(\.mainQueue) var mainQueue
        func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
          switch action {
          case .incrAndSquareLater:
            return .merge(
              EffectTask(value: .incrNow)
                .deferred(for: 2, scheduler: self.mainQueue), // compile error here 
              EffectTask(value: .squareNow)
                .deferred(for: 1, scheduler: self.mainQueue), // compile error here
              EffectTask(value: .squareNow)
                .deferred(for: 2, scheduler: self.mainQueue) // compile error here
            )
          case .incrNow:
            state += 1
            return .none
          case .squareNow:
            state *= state
            return .none
          }
        }
      }
   // (...)
}
```

Failing with the following error message:

> Argument type 'AnySchedulerOf<DispatchQueue>' (aka 'AnyScheduler<DispatchQueue.SchedulerTimeType, DispatchQueue.SchedulerOptions>') does not conform to expected type 'DateScheduler'
)

Possible solutions:

1. Creating/keeping a ReactiveSwift "version" of `MainQueue.swift` shadowing/redefining  the `\.mainQueue` property works but only when used inside the module where it is defined (e.g. defining it in ComposableArchitectureTests builds tests, but defining it inside ComposableArchitecture isn't enough for ComposableArchitectureTests to build). This also imposes an undesirable burden on library users while not scaling in multi module set ups.

2. We could work around it by defining a dependency having a different name with the `DateScheduler` type (e.g. `\.mainScheduler`) but we would be sacrificing the symmetry we've been trying to keep in regards to upstream TCA. 

3. I also tried making `AnySchedulerOf<DispatchQueue>` conform to ReactiveSwift's `DateScheduler` which could work but sadly ReactiveSwift's `Scheduler: AnyObject` meaning  another no-go with `AnySchedulerOf<DispatchQueue>` being a `struct`. We could always try to lift that restriction in ReactiveSwift though... 🤔  Linux support has to be ensured via `DateScheduler` anyway since it doesn't support Combine.

Hope this made it a bit clearer. Do you have any other suggestion? I might be missing something or simply _holding it wrong™_ 😅 

> And if this problem really is insurmountable without removing Combine schedulers, then you may have to fork this library to unblock you.

Yeah this is always an option, but I would like to avoid it if possible as it creates a new set of problems and increases the gap between both TCA forks 😐

**Reply by tgrapperon** [reply]
I'm maintaining [a companion library](https://github.com/tgrapperon/swift-dependencies-additions) that vends many additional dependencies, and this increases the chances that one or more dependencies from this library will "keypath-squat" the name that some end-users would like to use themselves.

So I'm contemplating annotating a few (or all) theses built-in dependencies with `@_disfavoredOverload`. According to my tests, it seems to work transitively: if `A` defines `d` disfavored, `B` defines `d` normally, and `C` depends on `A` and `B`, the definition of `d` from `B` seems to be reliably picked without ambiguity. I can confirm that `@_disfavoredOverload` is effective because the invocation becomes ambiguous if I remove it. Please also note that any local "override" of `d` defined in `C` will always be favored.

So one solution may be to annotate `DependencyValues.mainQueue` as `@_disfavoredOverload`. I don't know if it'll work reliably, but this is maybe something you can experiment with relatively simply on a fork of `swift-dependencies`.

**Reply by mbrandonw** [reply]
Hi @p4checo, thanks for the extra info.

I'd be curious if @tgrapperon's suggestion works for you in a fork of this library, and in fact dealing with "key path squatting" may be something we have to worry about in general, but another option that comes to mind is simply renaming your dependency key path. Perhaps `mainQueueScheduler`. It would be a breaking change for your library, but at least very easily fixable. Also, you can always use the shorter `mainQueue` naming for the variable:

```swift
@Dependency(\.mainQueueScheduler) var mainQueue
```

**Reply by p4checo** [reply]
hi @tgrapperon and @mbrandonw, thanks both!

> So one solution may be to annotate `DependencyValues.mainQueue` as `@_disfavoredOverload`. I don't know if it'll work reliably, but this is maybe something you can experiment with relatively simply on a fork of `swift-dependencies`.

I tested it in a fork of this library and can confirm that your suggestion works! Adding `@_disabledOverload` in `Dependencies` causes any overloaded property defined in `ComposableArchitecture` to be picked up correctly in a 3rd module, e.g. `ComposableArchitectureTests`, confirming your analysis/tests.

> I'd be curious if @tgrapperon's suggestion works for you in a fork of this library, and in fact dealing with "key path squatting" may be something we have to worry about in general

Yes, I think this can possibly be a problem that will become common as more people use this library and define their own dependencies using these "common" names. We already have two scenarios (me and @tgrapperon) and the library is just a couple weeks old 😆

I didn't think of `@_disfavoredOverload`, but it's a really good idea and solution for the "key path squatting" problem we're now facing, where we only have 2 "competing" external modules. I think this solution won't scale besides 2 external modules as ambiguity will arise if there's not at least one overload that's not `@_disfavoredOverload` or there's one defined in the current module... 🤔 

Anyway, in practice that should be quite uncommon and if it happens users can always use a different property name to easily work around this. That being said, IMO `@_disfavoredOverload` seems like a pretty innocuous solution for most cases. What do you say @mbrandonw?

>  but another option that comes to mind is simply renaming your dependency key path. Perhaps `\.mainQueueScheduler`. It would be a breaking change for your library, but at least very easily fixable. Also, you can always use the shorter `mainQueue` naming for the variable

Yes, this is essentially the solution 2 mentioned in my previous comment. The `\.mainQueueScheduler` name and idea of keeping the `mainQueue` naming in the "local" properties at the point of usage are nice though!

**Reply by tgrapperon** [reply]
@p4checo Just to clarify, the key path squatting was mostly a problem that my own library will create, but the same solution I'm planning to implement to mitigate this problem can be implemented in `swift-dependencies` too.

I don't think that `@_disfavoredOverload` should be something that everyone should consider, because as you remarked, it doesn't scale well. So it should probably only concern "base" level libraries, which main purpose is to provide these dependencies. I don't think there are many cases either for constantly redefining keypaths. In my case, I'm providing a `\.userDefaults` dependency, but I'm imagining a case where the user would like to build an abstraction over this dependency, but still want to call it `\.userDefaults`, because it is nicer. Your case is another example of the issue. But I don't think there are many ways to redefine `\.uuid` for example.

Furthermore, some escape routes will probably need to be provided by these "base" libraries to allow end-users (or even "middleware" libraries) to interact with the original dependency without ambiguity if needed. In my library, I was planning to re-expose dependencies behind a namespace (in my case, I was going to go `\.additions.userDefaults` for example, but we can maybe agree on a common pattern to gather these routes as extensions of a common type. For example `\.package.dependencies.mainQueue`, `\.package.dependenciesAdditions.userDefaults`, etc. You can also imagine re-exposing your `mainQueue` in `\.package.reactiveComposableArchitecture.mainQueue`. So at any point, the user could fall back to some unambiguous variant (for humans). But this kind of scheme requires some careful consideration before being made public.

I don't know what is the correct way to solve this issue. After a glance at the dependencies provided by `swift-dependencies`, I'm not seeing many that would make a lot of sense besides `\.calendar`, `\.timezone`, `\.locale` (I'm trying really hard to imagine cases), maybe `\.urlSession` and `\.openURL`, and of course `\.mainQueue`. But if the `@_disfavoredOverload` route is taken, I would be leaning toward only disfavoring declarations where someone has effectively raised a legitimate `key path squatting` issue, like in you did.

**Reply by pyrtsa** [reply]
While `@_disfavoredOverload` clearly works around the issue, it is not the right tool for the job [as it exists currently](https://github.com/apple/swift/blob/main/docs/ReferenceGuides/UnderscoredAttributes.md#_disfavoredoverload) (emphasis mine):

> Use `@_disfavoredOverload` to work around known bugs in the overload resolution rules that cannot be immediately fixed without a source break. **Don't use it to adjust overload resolution rules that are otherwise sensible but happen to produce undesirable results for your particular API**; it will likely be removed or made into a no-op eventually, and then you will be stuck with an overload set that cannot be made to function in the way you intend.

I think the valid ways forward are:
- renaming your key path (and face this issue again whenever bringing in libraries with conflicting dependency key paths).
- adding to `@propertyWrapper struct Dependency` an `init` overload which takes a `[Test]DependencyKey.Type` as its argument the same way the subscript access works on `DependencyValues`… and making your `DependencyKey`s `public`. The alternative property wrapper call site becomes: `@Dependency(MainQueueKey.self) var mainQueue` allowing for ordinary module qualification syntax for the parameter.

**Reply by mbrandonw** [reply]
I tend to agree with @pyrtsa and think we should avoid `@_disfavoredOverload` just to help out libraries that build upon this library. I think reactiveswift-composable-architecture renaming their dependency is the best long term solution, but also @pyrtsa's idea of having an escape hatch for unavoidable conflicts sounds good to me. @stephencelis and I will chat about that soon.

**Reply by tgrapperon** [reply]
I also concede that it would be exploiting the feature beyond its intended scope. If `Dependency` has an init with a `DependencyKey` type argument, end-users will be able to define their own overloads locally. It will only work at the module level though, but if they don't modularize their apps, it could work relatively seamlessly for them.
I have more liberty with my library, and I still have a few path to explore to solve this problem, especially considering a hypothetical `@Dependency(DependencyKey.Type)`.
I also hope that the day they remove `@_disfavoredOverload` (if it ever comes), we will have more tools at hand to implement this functionality.

**Reply by davdroman** [reply]
Proper module qualification should probably exist before Apple removes `@_disfavoredOverload`. I don't really see any other way they could justify removing it otherwise. Then it'd make sense to have `@Dependency(ReactiveSwift::MainQueue.self)` (hypothetical syntax).

Though at that point you might as well use namespaced keypaths e.g. `@Dependency(\.reactiveSwift.mainQueue)`.

**Reply by p4checo** [reply]
> @p4checo Just to clarify, the key path squatting was mostly a problem that my own library will create, but the same solution I'm planning to implement to mitigate this problem can be implemented in swift-dependencies too.

Oh I see. Thanks @tgrapperon for clarifying as I thought it was the other way around. 

Your idea of disambiguating via key path namespacing is interesting, but expecting everyone to follow a pattern/idiom might not be easy 😄

> While @_disfavoredOverload clearly works around the issue, it is not the right tool for the job [as it exists currently](https://github.com/apple/swift/blob/main/docs/ReferenceGuides/UnderscoredAttributes.md#_disfavoredoverload) (emphasis mine):

This is a fair point. While I don't expect this to be a common issue, I agree we should avoid relying on these "unofficial"/"unstable" attributes, especially in scenarios they haven't even been designed for (which seems to be the case).

I agree that your suggestion of allowing explicit type path qualification in the `Dependency` property wrapper is much more robust and should scale well for most or even all multi module ambiguity scenarios.

> I think reactiveswift-composable-architecture renaming their dependency is the best long term solution, but also @pyrtsa's idea of having an escape hatch for unavoidable conflicts sounds good to me. @stephencelis and I will chat about that soon.

I agree. Will follow your previous naming suggestion as I believe it achieves a nice balance. Hopefully we should see schedulers being used less and less in TCA soon 😁 

Thanks all for your inputs! 🙏🏼

**Reply by mbrandonw** [reply]
Hi @p4checo, thanks for being understanding and for adapting. It means a lot!

If this kind of thing continues to be an issue we will consider @pyrtsa's suggestion of an escape hatch to access dependencies by key, but for now we will keep things as-is.
